src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs

```diff
@@ -652,31 +653,39 @@ private void WritePropertyAccess(CodeRenderingContext context, ComponentAttribut
 
         context.CodeWriter.Write(DesignTimeVariable);
         context.CodeWriter.Write(" = ");
-        context.CodeWriter.Write("nameof(");
 
         if (componentNode.TypeInferenceNode == null)
         {
-            context.CodeWriter.Write("global::");
+            context.CodeWriter.Write("((global::");
             context.CodeWriter.Write(componentNode.Component.GetTypeNamespace());
             context.CodeWriter.Write(".");
             context.CodeWriter.Write(componentNode.Component.GetTypeNameIdentifier());
             if (componentNode.Component.IsGenericTypedComponent())
             {
+                // If there are generic type components, but no type inference node, then it means
+                // the user specified the type parameters, so we can use them directly
                 context.CodeWriter.Write("<");
-                var typeArgumentCount = componentNode.Component.GetTypeParameters().Count();
-                for (var i = 0; i < typeArgumentCount; i++)
+
+                var i = 0;
+                foreach (var typeArgumentNode in componentNode.Children.OfType<ComponentTypeArgumentIntermediateNode>())
                 {
-                    if (i > 0)
+                    if (i++ > 0)
+                    {
+                        context.CodeWriter.Write(", ");
+                    }
+
+                    foreach (var token in typeArgumentNode.Children.OfType<IntermediateToken>())
                     {
-                        context.CodeWriter.Write(",");
+                        // As per WriteComponentTypeArgument, we expect every token to be C#, but check just in case
+                        if (token.IsCSharp)
+                        {
+                            context.CodeWriter.Write(token.Content);
+                        }
                     }
-                    // We need to specify a type for the generic type parameters, but it doesn't matter which one, as we're just
-                    // getting the property name from nameof. If nameof ever supports open generics (ie, nameof(List<>.Count) then we
-                    // can remove this.
-                    context.CodeWriter.Write("string");
                 }
                 context.CodeWriter.Write(">");
             }
+            context.CodeWriter.Write(")null)");
```

Comment:
`default`, not `null`. `null` does not work if the type is a struct.

Reply:
Yep, should have thought of that. Will change this just to be careful, though being a component type I think it will be fine anyway.